core/src/main/java/org/apache/brooklyn/core/objs/ConstraintSerialization.java

```diff
@@ -0,0 +1,369 @@
+        public PredicateSerializationRuleAdder<T> equivalentPredicates(Predicate<?> ...equivs) {
+            for (Predicate<?> equiv: equivs) equivalentPredicateSamples.add(equiv);
+            return this;
+        }
+
+        public void add(ConstraintSerialization constraintSerialization) {
+            this.serialization = constraintSerialization;
+            if (predicateSample==null) predicateSample = constructor.apply(constructorSampleInput);
+            String toStringName = Strings.removeAfter(Preconditions.checkNotNull(predicateSample, "sample or sampleArg must be supplied").toString(), "(", false);
+            if (preferredName==null) {
+                preferredName = toStringName;
+            } else {
+                constraintSerialization.predicateToStringToPreferredName.put(preferredName, preferredName);
+            }
+            constraintSerialization.predicateToStringToPreferredName.put(toStringName, preferredName);
+            
+            for (String equiv: equivalentNames) {
+                constraintSerialization.predicateToStringToPreferredName.put(equiv, preferredName);
+            }
+            
+            constraintSerialization.predicatePreferredNameToConstructor.put(preferredName, constructor.compose(constructorArgsFromList));
+            
+            for (Predicate<?> equiv: equivalentPredicateSamples) {
+                String equivToStringName = Strings.removeAfter(equiv.toString(), "(", false);                
+                constraintSerialization.predicateToStringToPreferredName.put(equivToStringName, preferredName);
+            }
+        }
+    }
+    
+    private static String GROUP(String in) { return "("+in+")"; }
+    private static String NOT_CHARSET(String ...in) { return "[^"+Strings.join(in, "")+"]"; }
+    private static String OR_GROUP(String ...in) { return GROUP(Strings.join(in, "|")); }
+    private static String ZERO_OR_MORE(String in) { return in + "*"; }
+    
+    private static String DOUBLE_QUOTED_STRING = "\""+GROUP(ZERO_OR_MORE(OR_GROUP(NOT_CHARSET("\\", "\""), "\\.")))+"\"";
+    private static String SINGLE_QUOTED_STRING = "\'"+GROUP(ZERO_OR_MORE(OR_GROUP(NOT_CHARSET("\\", "\'"), "\\.")))+"\'";
+    
+    private static String PREDICATE = "[A-Za-z0-9_\\-\\.]+";
+    
+    private static Pattern PATTERN_START_WITH_QUOTED_STRING = Pattern.compile("^"+OR_GROUP(DOUBLE_QUOTED_STRING, SINGLE_QUOTED_STRING));
+    private static Pattern PATTERN_START_WITH_PREDICATE = Pattern.compile("^"+GROUP(PREDICATE));
+
+    {
+        init();
+    }
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private void init() {
+        PredicateSerializationRuleAdder.predicateListConstructor((o) -> ConfigConstraints.required()).
+            equivalentPredicates(Predicates.notNull(), StringPredicates.isNonBlank()).add(this);
+
+        PredicateSerializationRuleAdder.predicateListConstructor((o) -> Predicates.or((Iterable)o)).preferredName("any").equivalentNames("or").add(this);
```

Comment:
if we just have _one way_ then it _has_ to be the `toString()` which is ugly to work with -- or we have a different parser for the `toString()`.  given that we have `Predicates.or` and `any`, it would seem natural to have `or` as a similar word.  the disadvantages of having multiple ways is much reduced when one way is explicitly preferred, esp if we are able to convert things to a canonical form which we do here.  if you really don't like it i could drop `any` and `all`, just use `or` and `and` but don't think it's worth using the `toString()` or insisting on a separate parser to handle the conversion from the `toString()`.